backends/oauth2.go

```diff
@@ -0,0 +1,335 @@
+	}
+
+	if o.cacheIsValid(&cache) {
+		log.Debugf("using cached userinfo to authorize")
+	} else {
+		log.Debugf("update userinfo using authorization server %s", o.userInfoURL)
+
+		info, err := o.getUserInfo(cache.client)
+
+		if err != nil {
+			log.Errorf("Failed to receive UserInfo for user %s: %s", username, err)
+			return false, err
+		}
+
+		cache.superuser = info.MQTT.Superuser
+		cache.readTopics = info.MQTT.Topics.Read
+		cache.writeTopics = info.MQTT.Topics.Write
+		cache.updatedAt = time.Now()
+	}
+
+	log.Debugf("  user is superuser: %t", cache.superuser)
+	log.Debugf("  topics with read permission %s", cache.readTopics)
+	log.Debugf("  topics with write permission %s", cache.writeTopics)
+
+	res := o.checkAccessToTopic(topic, acc, &cache, username, clientid)
+	log.Debugf("ACL check was %t", res)
+	return res, nil
+}
+
+func (o Oauth2) GetName() string {
+	return "OAuth Plugin " + o.version
+}
+
+func (o Oauth2) Halt() {
+	// Do whatever cleanup is needed.
+}
+
+func (o Oauth2) createUserWithCredentials(username, password, clientid string) (bool, error) {
+	o.clientcredentialsConfig = go_clientcredentials.Config{
+		ClientID:     username,
+		ClientSecret: password,
+		TokenURL:     o.tokenUrl,
+	}
+
+	token, err := o.clientcredentialsConfig.Token(context.Background())
+
+	if err != nil {
+		log.Println(err)
+		return false, err
+	}
+
+	clientcredentialsClient := o.clientcredentialsConfig.Client(context.Background())
+
+	o.userCache[username] = userState{
+		username:  username,
+		superuser: false,
+		createdAt: time.Now(),
+		updatedAt: time.Unix(0, 0),
+		client:    clientcredentialsClient,
+		token:     token,
+	}
+
+	return true, err
+}
+
+func (o Oauth2) createUserWithToken(accessToken, clientid string) (bool, error) {
+	token := &go_oauth2.Token{
+		AccessToken: accessToken,
+		TokenType:   "Bearer",
+	}
+
+	client := o.oauth2Config.Client(context.Background(), token)
+
+	info, err := o.getUserInfo(client)
+
+	if err != nil {
+		log.Println(err)
+		return false, err
+	}
+
+	o.userCache[accessToken] = userState{
+		username:    accessToken,
+		superuser:   info.MQTT.Superuser,
+		createdAt:   time.Now(),
+		updatedAt:   time.Now(),
+		readTopics:  info.MQTT.Topics.Read,
+		writeTopics: info.MQTT.Topics.Write,
+		client:      client,
+		token:       token,
+	}
+
+	return true, err
+}
+
+func (o Oauth2) getUserInfo(client *http.Client) (*UserInfo, error) {
+	req, _ := http.NewRequest("GET", o.userInfoURL, nil)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	info := UserInfo{}
+
+	err = json.NewDecoder(resp.Body).Decode(&info)
+	if err != nil {
+		return nil, err
+	}
+
+	return &info, nil
+}
+
+func (o Oauth2) checkAccessToTopic(topic string, acc int32, cache *userState, username string, clientid string) bool {
+	log.Debugf("Check for acl level %d", acc)
+
+	// check read access
+	if acc == 1 || acc == 4 {
```

Comment:
Is there a reason not to allow for proper subscribe and read access and use them interchangeably?

Reply:
I split them up. But for now I cannot distinguish between those two because the server only responds with the pattern

```
{
    "sub": "<username>",
    "mqtt": {
        "topics": {
            "read": ["/test/#", "/test/read/#],
            "write": ["/test/#","/test/write/#]
        },
        "superuser": false
    }
}
```

I could expand the server response to distinguish between `MOSQ_ACL_SUBSCRIBE` and `MOSQ_ACL_WRITE` (or even between all `MOSQ_ACL_NONE`, `MOSQ_ACL_READ`, `MOSQ_ACL_WRITE`, `MOSQ_ACL_READWRITE`, `MOSQ_ACL_SUBSCRIBE`, `MOSQ_ACL_DENY`). But for now this simple version is totally sufficient for me.

Reply:
Yeah, in order to merge this we'll need a general solution, it can be tied to a specific use case.

Reply:
Ok, then I would expect a server reponse of the form

```
{
    "sub": "<username>",
    "mqtt": {
        "topics": {
            "read": ["/test/#"],
            "write": ["/test/#"],
            "subscribe": ["/subscribe/#"],
            "deny": ["/deny0","/deny/#"],
        },
        "superuser": false
    }
}
```

Which should be enough to check for all access levels `MOSQ_ACL_NONE`, `MOSQ_ACL_READ`, `MOSQ_ACL_WRITE`, `MOSQ_ACL_READWRITE`, `MOSQ_ACL_SUBSCRIBE`, `MOSQ_ACL_DENY`.

Reply:
Implemented in [commit](https://github.com/iegomez/mosquitto-go-auth/pull/210/commits/b979ff9e4eccd1ad1250c3222ec33e1cc7fb3342).